hcipy/coronagraphy/perfect_coronagraph.py

```diff
@@ -4,132 +4,132 @@
-		modes, which can be used to design perfect coronagraphs that are insensitive to stellar
-		radius [Guyon2006]_. If it is None, all modes are completely suppressed.
-	'''
-	def __init__(self, aperture, order=2, coeffs=None):
-		self.pupil_grid = aperture.grid
-		modes = []
-
-		if coeffs is not None:
-			order = int(2 * np.ceil(0.5 * (np.sqrt(8 * len(coeffs) + 1) - 1)))
-			self.coeffs = coeffs
-		else:
-			self.coeffs = np.ones(int(order * (order / 2 + 1) / 4))
-
-		for i in range(order // 2):
-			for j in range(i + 1):
-				modes.append(aperture * self.pupil_grid.x**j * self.pupil_grid.y**(i - j))
-
-		self.mode_basis = ModeBasis(modes).orthogonalized
-
-		self.transformation = self.mode_basis.transformation_matrix
-		self.transformation_inverse = inverse_truncated(self.transformation, 1e-6)
-
-	def forward(self, wavefront):
-		'''Propagate the wavefront through the perfect coronagraph.
-
-		Parameters
-		----------
-		wavefront : Wavefront
-			The incoming wavefront (in the pupil plane)
-
-		Returns
-		-------
-		Wavefront
-			The post-coronagraphic wavefront (in the pupil plane).
-		'''
-		wf = wavefront.copy()
-
-		# Project the wavefront onto the pupil modes and attenuate by coeffs, then transform back to the pupil.
-		# This is done for each polarization separately.
-		correction = np.einsum(
-			'kj,j,ji,...i->...k',
-			self.transformation,
-			self.coeffs,
-			self.transformation_inverse,
-			wf.electric_field,
-			optimize='optimal'
-		)
-
-		# Then subtract this from the original wavefront to compute the post-coronagraphic field.
-		wf.electric_field -= correction
-
-		return wf
-
-	def backward(self, wavefront):
-		'''Propagate the wavefront backwards through the perfect coronagraph.
-
-		This method behaves the same as the forward propagation.
-
-		Parameters
-		----------
-		wavefront : Wavefront
-			The incoming wavefront (in the pupil plane)
-
-		Returns
-		-------
-		Wavefront
-			The post-coronagraphic wavefront (in the pupil plane).
-		'''
-		return self.forward(wavefront)
-
-	def get_transformation_matrix_forward(self, wavelength=1):
-		'''Get the forward propagation transformation matrix.
-
-		Parameters
-		----------
-		wavelength : scalar
-			The wavelength at which to calculate the transformation matrix.
-
-		Returns
-		-------
-		ndarray
-			The forward transformation_matrix.
-		'''
-		return np.eye(self.pupil_grid.size) - self.transformation.dot(self.coeffs * self.transformation_inverse)
-
-	def get_transformation_matrix_backward(self, wavelength=1):
-		'''Get the backwards propagation transformation matrix.
-
-		This method behaves the same as the forward propagation.
-
-		Parameters
-		----------
-		wavelength : scalar
-			The wavelength at which to calculate the transformation matrix.
-
-		Returns
-		-------
-		ndarray
-			The backward transformation_matrix.
-		'''
-		return self.get_transformation_matrix_forward(wavelength)
+    r'''A perfect coronagraph for a certain aperture and order.
```

Comment:
Completely unrelated to this PR, but is this `r` here on purpose? I found one more further below.

Reply:
The `r` is for raw string, which escapes some of the math in this docstring.